dashboard/app/controllers/xhr_proxy_controller.rb

```diff
@@ -0,0 +1,47 @@
+# A controller for proxying web requests to 3rd party APIs. This protects
+# applab users from XSS attacks, complies with the Same Origin Policy,
+# and prevents http/https mismatch warnings in IE.
+#
+# Responses are cached for one minute since many 3rd party APIs serve data
+# which changes frequently such as news or sports scores.
+#
+# To reduce the likelihood of abuse, we only proxy content with an allowed
+# whitelist of media types. We will need to monitor usage to detect
+# abuse and potentially add other abuse prevention measures.
+
+require 'set'
+
+class XhrProxyController < ApplicationController
+  include ProxyHelper
+
+  ALLOWED_CONTENT_TYPES = Set.new(['application/json'])
+
+  # How long the content is allowed to be cached
+  EXPIRY_TIME = 1.minute
+
+  # Return the proxied api at the given URL.
+  def get
+    channel_id = params[:c]
+    url = params[:u]
+
+    begin
+      owner_storage_id, _ = storage_decrypt_channel_id channel_id
+    rescue Exception => e
+      raise "XhrProxyController request with invalid channel_id: '#{channel_id}' for url: '#{url}' exception: #{e.message}"
```

Comment:
It would be better to return a HTTP 403 (denied) for an invalid token rather than a 500 (server error) since the request itself is invalid.

The error message should refer to an invalid token rather than an invalid channel id since that is an internal implementation detail that we don't want to expose to the user.